### PR TITLE
Deduplicate notifications

### LINKED ISSUE #1483: Duplicated notifications
### Current situation

Currently when you have the desktop client installed, the iOS app on the iPad, the Android app on the phone, your browser(s) open, you will receive call notifications multiple times in too many locations

### Proposed plan

We plan to reduce the notifications to only push to:
* A single desktop client type
* The mobile apps of the same type
* Browser not showing notifications if a desktop client has connected to the server in reasonable timeout range

Since the current desktop client simply pulls the API and the browsers too, we will add a new field there indicating for them if they should show notifications or not.

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #1483

### CODE REVIEW

File: lib/Service/ClientService.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+		$result = $query->executeQuery();
+		while ($row = $result->fetch()) {
+			\OC::$server->getLogger()->error(json_encode($row['name']));
+			if (preg_match('/ \(Talk Desktop Client - [A-Za-z ]+\)$/', $row['name'])) {
+				$result->closeCursor();
+				return true;
+			}
+		}
+		$result->closeCursor();
```
Comment (ShGKme) [inline lib/Service/ClientService.php:40]: From a perf perspective, wouldn't it be better to add this as a part of SQL query?
Aka `AND name LIKE ' %(Talk Desktop Client -%)'

Comment (nickvergessen) [reply lib/Service/ClientService.php:40]: That breaks index usage as far as I remember, and the result set afterwards is so small we are better off checking the few for your user afterwards.
But will check again which indexes are available, but I'm pretty sure name had none

File: lib/Controller/EndpointController.php [hunk 2]
```diff
@@ -247,12 +252,19 @@ protected function notificationToArray(int $notificationId, INotification $notif
 		];
 
 		if ($apiVersion !== 'v1') {
+			if ($this->request->isUserAgent([IRequest::USER_AGENT_TALK_DESKTOP])) {
+				$shouldNotify = $notification->getApp() === 'spreed';
+			} else {
+				$shouldNotify = !$hasActiveTalkDesktop || $notification->getApp() !== 'spreed';
+			}
```
Comment (ShGKme) [inline lib/Controller/EndpointController.php:259]: I have a 500 response here

```
Undefined constant OCP\IRequest::USER_AGENT_TALK_DESKTOP in file '/var/www/html/apps/notifications/lib/Controller/EndpointController.php' line 255
```

Does it require a server update? In this case, shouldn't we check it here for compatibility?

Comment (nickvergessen) [reply lib/Controller/EndpointController.php:259]: git pull server yeah. It as added yesterday

**Review by ShGKme — APPROVED** [review]
Works 👀

**Review by vitormattos — CHANGES_REQUESTED** [review]

File: lib/Service/ClientService.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace OCA\Notifications\Service;
+
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\IDBConnection;
+use OCP\IRequest;
+
+class ClientService {
+	public const DESKTOP_CLIENT_TIMEOUT = 120;
```
[the remark this answers is not in the document]
Comment (nickvergessen) [reply lib/Service/ClientService.php:33]: It's including a technical limitation. The authtoken's last_activity column is only updated after 60 seconds. So let's assume this is 59 seconds (worst case), and then the desktop client is only pinging every 30 seconds for messages, but the request could take 1-2 seconds over the network, leaving us with at least 91+ seconds. And some more buffer is better save to be assumed, so we decided on 120 seconds.

**Review by vitormattos — APPROVED** [review]

**Pull request merged**